File: homework02/sudoku.go
package main

import (
	"fmt"
	"io/ioutil"
	"math/rand"
	"path/filepath"
	"sort"
	"time"
)

func readSudoku(filename string) ([][]byte, error) {
	data, err := ioutil.ReadFile(filename)
	if err != nil {
		return nil, err
	}
	grid := group(filter(data), 9)
	return grid, nil
}

func filter(values []byte) []byte {
	filtered_values := make([]byte, 0)
	for _, v := range values {
		if (v >= '1' && v <= '9') || v == '.' {
			filtered_values = append(filtered_values, v)
		}
	}
	return filtered_values
}

func display(grid [][]byte) {
	for i := 0; i < len(grid); i++ {
		for j := 0; j < len(grid); j++ {
			fmt.Print(string(grid[i][j]))
		}
		fmt.Println()
	}
}

func group(values []byte, n int) [][]byte {
	var a [][]byte
	for i := 0; i < len(values); i += n {
		a = append(a, values[i:i+n])
	}
	return a
}

func getRow(grid [][]byte, row int) []byte {
	return grid[row]
}

func getCol(grid [][]byte, col int) []byte {
	var column []byte
	for i := 0; i < len(grid); i++ {
		column = append(column, grid[i][col])
	}
	return column
}

func getBlock(grid [][]byte, row int, col int) []byte {
	var block_val []byte
	var subrow, subcol int
	subrow = (row / 3) * 3
	subcol = (col / 3) * 3
	for i := 0; i < 3; i++ {
		for j := 0; j < 3; j++ {
			block_val = append(block_val, grid[subrow+i][subcol+j])
		}
	}
	return block_val
}

func findEmptyPosition(grid [][]byte) (int, int) {
	for row := 0; row < len(grid); row++ {
		for col := 0; col < len(grid); col++ {
			if grid[row][col] == '.' {
				return row, col
			}
		}

	}
	return -1, -1
}

func contains(values []byte, search byte) bool {
	for _, v := range values {
		if v == search {
			return true
		}
	}
	return false
}

func findPossibleValues(grid [][]byte, row int, col int) []byte {
	var Values []byte
	a := []byte("123456789")
	row1 := getRow(grid, row)
	col1 := getCol(grid, col)
	block1 := getBlock(grid, row, col)
	for i := 0; i < 9; i++ {
		flag := false
		for j := 0; j < 9; j++ {
			if (a[i] == row1[j]) || (a[i] == col1[j]) || (a[i] == block1[j]) {
				flag = true
			}
		}
		if flag == false {
			Values = append(Values, a[i])
		}
	}
	return Values
}

func solve(grid [][]byte) ([][]byte, bool) {
	row, col := findEmptyPosition(grid)
	if (row == -1) {
		return grid, true
	}
	for _, value := range (findPossibleValues(grid, row, col)) {
		grid[row][col] = value
		solution, answ := solve(grid)
		if answ {
			return solution, answ
		}
	}
	grid[row][col] = '.'
	var none [][]byte
	return none, false
}

func checkSolution(grid [][]byte) bool {
	var int_val []int
	a := [9]int{1, 2, 3, 4, 5, 6, 7, 8, 9}

	for row := 0; row < len(grid); row++ {
		var row_values []byte
		row_values = getRow(grid, row)
		for i := 0; i < len(row_values); i++ {
			int_val = append(int_val, int(row_values[i]))
		}
		sort.Ints(int_val)
		for i := 0; i < len(int_val); i++ {
			if int_val[i] != a[i] {
				return false
			}
		}
	}
	int_val = nil
	for col := 0; col < len(grid); col++ {
		var col_values []byte
		col_values = getCol(grid, col)
		for i := 0; i < len(col_values); i++ {
			int_val = append(int_val, int(col_values[i]))
		}
		sort.Ints(int_val)
		for i := 0; i < len(int_val); i++ {
			if int_val[i] != a[i] {
				return false
			}
		}
	}
	int_val = nil
	for row := 0; row <= 6; row += 3 {
		for col := 0; col <= 6; col += 3 {
			var block []byte
			block = getBlock(grid, row, col)
			for i := 0; i < len(block); i++ {
				int_val = append(int_val, int(block[i]))
			}
			sort.Ints(int_val)
			for i := 0; i < len(int_val); i++ {
				if int_val[i] != a[i] {
					return false
				}
			}
		}
	}
	return true
}

func generateSudoku(N int) [][]byte {
	var matrix [][]byte
	a := []byte(".........")
	for i := 0; i < 9; i++ {
		matrix = append(matrix, a)
	}
	matrix, _ = solve(matrix)
	if N > 81 {
		N = 0
	} else {
		N = 81 - N
	}
	rand.Seed(time.Now().UnixNano())
	for N > 0 {
		row := rand.Intn(8)
		col := rand.Intn(8)
		if matrix[row][col] != '.' {
			matrix[row][col] = '.'
		}
		N -= 1
	}
	return matrix
}

func main() {
	puzzles, err := filepath.Glob("puzzle*.txt")
	if err != nil {
		fmt.Printf("Could not find any puzzles")
		return
	}
	for _, fname := range puzzles {
		go func(fname string) {
			grid, _ := readSudoku(fname)
			solution, _ := solve(grid)
			fmt.Println("Solution for", fname)
			display(solution)
		}(fname)
	}
	var input string
	fmt.Scanln(&input)
}

File: homework02/sudoku.py
import multiprocessing
import random
from typing import Union


def read_sudoku(filename: str) -> list:
    """ Прочитать Судоку из указанного файла """
    digits = [c for c in open(filename).read() if c in '123456789.']
    grid = group(digits, 9)
    return grid


def display(values: list) -> None:
    """Вывод Судоку """
    width = 2
    line = '+'.join(['-' * (width * 3)] * 3)
    for row in range(9):
        print(''.join(values[row][col].center(width) + ('|' if str(col) in '25' else '') for col in range(9)))
        if str(row) in '25':
            print(line)
    print()


def group(values: list, n: int) -> list:
    """
    Сгруппировать значения values в список, состоящий из списков по n элементов
    >>> group([1,2,3,4], 2)
    [[1, 2], [3, 4]]
    >>> group([1,2,3,4,5,6,7,8,9], 3)
    [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
    """
    return [values[i:i + n] for i in range(0, len(values), n)]


def get_row(values: list, pos: tuple) -> list:
    """ Возвращает все значения для номера строки, указанной в pos
    >>> get_row([['1', '2', '.'], ['4', '5', '6'], ['7', '8', '9']], (0, 0))
    ['1', '2', '.']
    >>> get_row([['1', '2', '3'], ['4', '.', '6'], ['7', '8', '9']], (1, 0))
    ['4', '.', '6']
    >>> get_row([['1', '2', '3'], ['4', '5', '6'], ['.', '8', '9']], (2, 0))
    ['.', '8', '9']
    """
    row, _ = pos
    return values[row]


def get_col(values: list, pos: tuple) -> list:
    """ Возвращает все значения для номера столбца, указанного в pos
    >>> get_col([['1', '2', '.'], ['4', '5', '6'], ['7', '8', '9']], (0, 0))
    ['1', '4', '7']
    >>> get_col([['1', '2', '3'], ['4', '.', '6'], ['7', '8', '9']], (0, 1))
    ['2', '.', '8']
    >>> get_col([['1', '2', '3'], ['4', '5', '6'], ['.', '8', '9']], (0, 2))
    ['3', '6', '9']
    """
    _, column = pos
    return [values[i][column] for i in range(len(values))]


def get_block(values: list, pos: tuple) -> list:
    """ Возвращает все значения из квадрата, в который попадает позиция pos
    >>> grid = read_sudoku('puzzle1.txt')
    >>> get_block(grid, (0, 1))
    ['5', '3', '.', '6', '.', '.', '.', '9', '8']
    >>> get_block(grid, (4, 7))
    ['.', '.', '3', '.', '.', '1', '.', '.', '6']
    >>> get_block(grid, (8, 8))
    ['2', '8', '.', '.', '.', '5', '.', '7', '9']
    """
    row, column = pos
    subrow = row // 3 * 3
    subcol = column // 3 * 3
    block_values = []
    for i in range(3):
        for j in range(3):
            block_values.append(values[subrow + i][subcol + j])
    return block_values


def find_empty_positions(grid: list) -> Union[None, tuple]:
    """ Найти первую свободную позицию в пазле
    >>> find_empty_positions([['1', '2', '.'], ['4', '5', '6'], ['7', '8', '9']])
    (0, 2)
    >>> find_empty_positions([['1', '2', '3'], ['4', '.', '6'], ['7', '8', '9']])
    (1, 1)
    >>> find_empty_positions([['1', '2', '3'], ['4', '5', '6'], ['.', '8', '9']])
    (2, 0)
    """
    for row in range(len(grid)):
        for column in range(len(grid)):
            if grid[row][column] == '.':
                return (row, column)
    return None


def find_possible_values(grid: list, pos: tuple) -> set:
    """ Вернуть множество возможных значения для указанной позиции
    >>> grid = read_sudoku('puzzle1.txt')
    >>> values = find_possible_values(grid, (0,2))
    >>> values == {'1', '2', '4'}
    True
    >>> values = find_possible_values(grid, (4,7))
    >>> values == {'2', '5', '9'}
    True
    """
    possible_values = {'1', '2', '3', '4', '5', '6', '7', '8', '9'}
    return possible_values - set(get_row(grid, pos)) - set(get_col(grid, pos)) - set(get_block(grid, pos))


def solve(grid: list) -> Union[None, list]:
    """ Решение пазла, заданного в grid """
    """ Как решать Судоку?
        1. Найти свободную позицию
        2. Найти все возможные значения, которые могут находиться на этой позиции
        3. Для каждого возможного значения:
            3.1. Поместить это значение на эту позицию
            3.2. Продолжить решать оставшуюся часть пазла
    >>> grid = read_sudoku('puzzle1.txt')
    >>> solve(grid)
    [['5', '3', '4', '6', '7', '8', '9', '1', '2'], ['6', '7', '2', '1', '9', '5', '3', '4', '8'], ['1', '9', '8', '3', '4', '2', '5', '6', '7'], ['8', '5', '9', '7', '6', '1', '4', '2', '3'], ['4', '2', '6', '8', '5', '3', '7', '9', '1'], ['7', '1', '3', '9', '2', '4', '8', '5', '6'], ['9', '6', '1', '5', '3', '7', '2', '8', '4'], ['2', '8', '7', '4', '1', '9', '6', '3', '5'], ['3', '4', '5', '2', '8', '6', '1', '7', '9']]
    """
    pos = find_empty_positions(grid)
    if pos is None:
        return grid
    row, column = pos
    for dig in find_possible_values(grid, pos):
        grid[row][column] = dig
        solution = solve(grid)
        if solution:
            return solution
    grid[row][column] = '.'
    return None


def check_solution(solution: list) -> bool:
    """ Если решение solution верно, то вернуть True, в противном случае False """
    # TODO: Add doctests with bad puzzles
    nums = {'1', '2', '3', '4', '5', '6', '7', '8', '9'}

    for row in range(len(solution)):
        row_values = set(get_row(solution, (row, 0)))
        if row_values != nums:
            return False

    for column in range(len(solution)):
        column_values = set(get_row(solution, (0, column)))
        if column_values != nums:
            return False

    for row in (0, 3, 6):
        for column in (0, 3, 6):
            block = set(get_block(solution, (row, column)))
            if block != nums:
                return False

    return True


def generate_sudoku(N: int) -> list:
    """ Генерация судоку заполненного на N элементов
    >>> grid = generate_sudoku(40)
    >>> sum(1 for row in grid for e in row if e == '.')
    41
    >>> solution = solve(grid)
    >>> check_solution(solution)
    True
    >>> grid = generate_sudoku(1000)
    >>> sum(1 for row in grid for e in row if e == '.')
    0
    >>> solution = solve(grid)
    >>> check_solution(solution)
    True
    >>> grid = generate_sudoku(0)
    >>> sum(1 for row in grid for e in row if e == '.')
    81
    >>> solution = solve(grid)
    >>> check_solution(solution)
    True
    """
    matrix = solve([['.'] * 9 for _ in range(9)])
    N = 81 - min(81, max(0, N))
    while N > 0:
        row = random.randint(0, 8)
        column = random.randint(0, 8)
        if matrix[row][column] != '.':
            matrix[row][column] = '.'
            N -= 1
    return matrix


def run_solve(fname: str) -> None:
    grid = read_sudoku(fname)
    display(grid)
    solution = solve(grid)
    display(solution)


if __name__ == '__main__':
    for fname in ['puzzle1.txt', 'puzzle2.txt', 'puzzle3.txt']:
        p = multiprocessing.Process(target=run_solve, args=(fname,))
        p.start()